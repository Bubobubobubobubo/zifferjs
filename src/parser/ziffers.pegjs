{
  var nodeOptions = options.nodeOptions || {};
  var transform = options.transform || function(value) {};
  var DEFAULT_DURATIONS = options.defaultDurs;

  var Node = function(values) { 
    // Merge all values properties to this
    for (var key in values) {
      if(values[key] !== null) {
        this[key] = values[key];
      }
    }
    this.location = location();
    if(this.type === 'pitch') {
      // Merge all default options to the node if value is not set, null or undefined
      for (var key in nodeOptions) {
        if (this[key] === undefined || this[key] === null) {
          this[key] = nodeOptions[key];
        }
      }
      // Transform the node
      transform(this);
    }
  }

  var seed = 0;

   // console.log("OPTIONS:", options);
  
}

start = s:statement { return s.filter(a => a) }

// ----- Numbers -----

float = ("-"? [0-9]* "." [0-9]+ / "." [0-9]+) { return parseFloat(text()) }

int = "-"? [0-9]+ { return parseInt(text()); }

// ------------------ delimiters ---------------------------

ws "whitespace" = [ \n\r\t] { return undefined }
comma = ws "," ws
pipe = ws "|" ws
quote = '"' / "'"


durchar = [a-z] { return DEFAULT_DURATIONS[text()]; }

duration = durchar / float

statement = items

items = n:(repeat / list / item)+
{ return n }

list = "(" l:(items) ")"
{ return l.filter(a => a) }

item = v:(chord / pitch / ws / duration_change / list)
{ return v }

repeat = n:item ":" i:int
{ return new Node({type: 'repeat', value: n, times: i}) }

duration_change = dur:duration { 
  options.nodeOptions.duration = dur;
  return new Node({type: 'duration_change', duration: dur}) 
  }

pitch = dur:duration? i:int 
{ return new Node({type: 'pitch', duration: dur, value: i, pitch: parseInt(i)}) }

chord = left:pitch right:pitch+
{ return new Node({type: 'chord', pitches:[left].concat(right)}) }
