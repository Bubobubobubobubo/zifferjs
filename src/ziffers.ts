import { parse } from './parser/parser.ts';
import {DEFAULT_DURS, DEFAULT_OPTIONS} from './defaults.ts';
import { noteFromPc, midiToFreq } from './scale.ts';

interface Options {
    nodeOptions?: NodeOptions;
    transform?: Function
    defaultDurs?: {[key: string]: number}
}

interface NodeOptions {
    key?: string;
    scale?: string;
    duration?: number;
}

interface Node {
    type: string;
    value: string;
    location: Location;
    duration?: number;
    pitch?: number;
    pitches?: Node[];
    freq?: number;
    key?: string;
    scale?: string;
    note?: number;
    octave?: number;
    bend?: number;
}

interface RepeatNode extends Node {
    value: Node;
    times: number;
}

interface Location {
    end: Position;
    start: Position;
}

interface Position {
    offset: number;
    line: number;
    column: number;
}

export class Ziffers {
    values: (Node|RepeatNode)[];
    evaluated: Node[];
    options: Options;

    constructor(input: string, options: NodeOptions = {}) {
        console.log("Parsing:", input);
        // Merge options with default options
        options = {...DEFAULT_OPTIONS, ...options};
        this.options = {nodeOptions: options, transform: transform, defaultDurs: DEFAULT_DURS};

        // TODO: Parse scala workshop format

        try {
           this.values = parse(input, this.options);
           // TODO: Evaluate
           this.evaluate();
        } catch (ex: any) {
            console.log(ex);
            // Handle parsing error
            // [...]
            this.values = [];
            this.evaluated = [];
        }
    }

    evaluate() {
        // Evaluate
        this.evaluated = evaluateNodes(this.values);
    }

    pitches() {
        return this.evaluated.map((item: Node) => {
            return values.pitch;
        });
    }

}

const evaluateNodes = (nodes: Node[]) => {
    return nodes.map((node) => {
        if(node.type === 'repeat') {
            const times = node.times;
            // Duplicate values given times
            const repeated = [...Array(times)].map(() => node.value).flat(Infinity);
            return evaluateNodes(repeated);
        }
        return node;
    }).flat(Infinity);
}

// Function to transform the AST node on parse time
const transform = (node: Node) => {
    if(node.type === 'pitch') {
        const [note, bend] = noteFromPc(node.key!, node.pitch!, node.scale!);
        node.note = note;
        node.freq = midiToFreq(note);
        if(bend) node.bend = bend;
    }
} 

export const zparse = (input: string, options: NodeOptions ) => {
    return new Ziffers(input, options);
}