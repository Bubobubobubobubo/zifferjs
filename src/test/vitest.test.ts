import { assert, describe, expect, it } from 'vitest'
import { zparse } from '../ziffers.ts'

describe('main-tests', () => {
  it('parse', () => {
    expect(zparse('1',{scale: "minor"}).values.length).toEqual(1);
    expect(zparse('12 3 4 5').values.length).toEqual(4);
    expect(1 + 1).toEqual(2);
    expect(true).to.be.true;
  })

  it('durations', () => {
    expect(zparse('e 1').values[0].duration).toEqual(0.125);
    //assert.equal(Math.sqrt(4), 2)
  })

  it('snapshot', () => {
    expect({ foo: 'bar' }).toMatchSnapshot()
  })

  it('print', () => {
    const list = [
      '1 2 3',
      '0.3 ((1):2):2 q4:5 q1 (1 2(3 4)):2']
    list.forEach((z) => {
      console.log(zparse(z,{}));
    })
  })
})