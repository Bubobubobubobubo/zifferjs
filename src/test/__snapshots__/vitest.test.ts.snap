// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`main-tests > snapshot 1`] = `
{
  "foo": "bar",
}
`;
